### PR TITLE
neural networks

### PR DESCRIPTION
**Pull request by Sami19944**
uses FANN to train neuralnetworks to drive the car

### CODE REVIEW

**Comment by Nils-Jung** [comment]
- Should the `.gitignore.swp` and `.gitignore.save` files be in there?
- You are downgrading the `vesc` submodule from b7eb0b (https://github.uconn.edu/ret15108/vesc, new) to 5127d6 (https://github.com/mit-racecar/vesc, old). Please fix that.

**Comment by marian42** [comment]
Since this implements an evolutionary algorithm, the package should be called something like "evolutionary driver".

File: ros_ws/src/autonomous/neuralnetwork/configs/init/latest.conf [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+train_error_function=1
+train_stop_function=0
+cascade_output_change_fraction=0.010000
+quickprop_decay=-0.000100
+quickprop_mu=1.750000
+rprop_increase_factor=1.200000
+rprop_decrease_factor=0.500000
+rprop_delta_min=0.000000
+rprop_delta_max=50.000000
+rprop_delta_zero=0.100000
```
Comment (marian42) [inline ros_ws/src/autonomous/neuralnetwork/configs/init/latest.conf]: Please check if you can remove the learning-related parameters since they will not be used.

**Comment by marian42** [comment]
Where are the activation functions configured and which activation functions are used?

**Comment by marian42** [comment]
I think that FANN is not a good choice of library for vector math. If this was done using pytorch or even numpy, the entirety of `ai_math` and `ai_utils` would not be needed. The operations implemented in these files are also available in eigen if you want to stay with C++.

File: ros_ws/launch/neuralnetwork.launch [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+
+    <arg name="mode_training" default="true" />
+
+    <node
+      respawn="true"
+      pkg="neuralnetwork"
+      type="ai_driver"
+      name="ai_driver"
+      output="screen" >
+      <param name="update_rate" type="double" value = "10"/> <!-- Hz-->
```
[the remark this answers is not in the document]
Comment (Sami19944) [reply ros_ws/launch/neuralnetwork.launch]: The update rate can be set to 60, but it has to be fixed. A variable update rate can mess with the input values.

File: ros_ws/src/autonomous/neuralnetwork/configs/init/latest.conf (hunk 1 above)
Comment (Sami19944) [reply ros_ws/src/autonomous/neuralnetwork/configs/init/latest.conf]: the save function isnt implemented by me, so i dont thinks its possible

File: ros_ws/src/autonomous/neuralnetwork/include/ai_util.h [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+                {
+                    continue;
+                }
+                vec.push_back(folder + "/" + name);
+            }
+            closedir(dir);
+        }
+        else
+        {
+            /* could not open directory */
```
[the remark this answers is not in the document]
Comment (Sami19944) [reply ros_ws/src/autonomous/neuralnetwork/include/ai_util.h]: > Where are the activation functions configured and which activation functions are used?

The default activation function is the sigmoid function. As I only need this one, no other is configured

**Comment by Sami19944** [comment]
> I think that FANN is not a good choice of library for vector math. If this was done using pytorch or even numpy, the entirety of `ai_math` and `ai_utils` would not be needed. The operations implemented in these files are also available in eigen if you want to stay with C++.

ai_math and ai_util are only miniture libaries and using bigger ones would be, in my view, overkill
